bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/models/v1/TableOfContentsImpl.java

```diff
@@ -0,0 +1,123 @@
+package com.adobe.cq.wcm.core.components.internal.models.v1;
+
+import com.adobe.cq.export.json.ExporterConstants;
+import com.adobe.cq.wcm.core.components.models.TableOfContents;
+import com.day.cq.wcm.api.designer.Style;
+import org.apache.sling.api.SlingHttpServletRequest;
+import org.apache.sling.models.annotations.Exporter;
+import org.apache.sling.models.annotations.Model;
+import org.apache.sling.models.annotations.injectorspecific.InjectionStrategy;
+import org.apache.sling.models.annotations.injectorspecific.ScriptVariable;
+import org.apache.sling.models.annotations.injectorspecific.Self;
+import org.apache.sling.models.annotations.injectorspecific.ValueMapValue;
+import org.jetbrains.annotations.Nullable;
+
+import javax.annotation.PostConstruct;
+import java.util.Arrays;
+
+@Model(
+    adaptables = SlingHttpServletRequest.class,
+    adapters = {
+        TableOfContents.class
+    },
+    resourceType = TableOfContentsImpl.RESOURCE_TYPE
+)
+@Exporter(name = ExporterConstants.SLING_MODEL_EXPORTER_NAME,
+    extensions = ExporterConstants.SLING_MODEL_EXTENSION)
+public class TableOfContentsImpl implements TableOfContents {
+
+    /**
+     * The resource type of the component associated with this Sling model.
+     */
+    public static final String RESOURCE_TYPE = "core/wcm/components/tableofcontents/v1/tableofcontents";
+
+    public static final String DEFAULT_LIST_TYPE = "unordered";
+    public static final Integer DEFAULT_START_LEVEL = 1;
+    public static final Integer DEFAULT_STOP_LEVEL = 6;
+
+    public static final String NO_RESTRICTION = "norestriction";
+
+    @Self(injectionStrategy = InjectionStrategy.REQUIRED)
+    private SlingHttpServletRequest slingHttpServletRequest;
+
+    @ScriptVariable
+    protected Style currentStyle;
+
+    @ValueMapValue(name = TableOfContents.PN_LIST_TYPE, injectionStrategy = InjectionStrategy.OPTIONAL)
+    @Nullable
+    private String listType;
+
+    @ValueMapValue(name = TableOfContents.PN_START_LEVEL, injectionStrategy = InjectionStrategy.OPTIONAL)
+    @Nullable
+    private Integer startLevel;
+
+    @ValueMapValue(name = TableOfContents.PN_STOP_LEVEL, injectionStrategy = InjectionStrategy.OPTIONAL)
+    @Nullable
+    private Integer stopLevel;
+
+    private String restrictListType;
+    private String restrictStartLevel;
+    private String restrictStopLevel;
+    private String[] includeClasses;
+    private String[] ignoreClasses;
+
+    @PostConstruct
+    private void initModel() {
+        restrictListType = currentStyle.get(PN_RESTRICT_LIST_TYPE, String.class);
+        restrictStartLevel = currentStyle.get(PN_RESTRICT_START_LEVEL, String.class);
+        restrictStopLevel = currentStyle.get(PN_RESTRICT_STOP_LEVEL, String.class);
+        includeClasses = currentStyle.get(PN_INCLUDE_CLASSES, String[].class);
+        ignoreClasses = currentStyle.get(PN_IGNORE_CLASSES, String[].class);
+        slingHttpServletRequest.setAttribute("contains-table-of-contents", true);
```

Comment:
I'd prefix with `cmp` to avoid clashes with custom attributes: `cmp:contains-table-of-contents`